istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -222,21 +225,116 @@ func createRemoteSecretFromTokenAndServer(tokenSecret *v1.Secret, clusterName, s
-	serviceAccount, err := kube.CoreV1().ServiceAccounts(saNamespace).Get(context.TODO(), saName, metav1.GetOptions{})
+func getServiceAccountSecret(client kube.ExtendedClient, opt RemoteSecretOptions) (*v1.Secret, error) {
+	// Create the service account if it doesn't exist.
+	serviceAccount, err := getOrCreateServiceAccount(client, opt)
 	if err != nil {
 		return nil, err
 	}
+
 	if len(serviceAccount.Secrets) != 1 {
 		return nil, fmt.Errorf("wrong number of secrets (%v) in serviceaccount %s/%s",
-			len(serviceAccount.Secrets), saNamespace, saName)
+			len(serviceAccount.Secrets), opt.Namespace, opt.ServiceAccountName)
 	}
+
 	secretName := serviceAccount.Secrets[0].Name
 	secretNamespace := serviceAccount.Secrets[0].Namespace
 	if secretNamespace == "" {
-		secretNamespace = saNamespace
+		secretNamespace = opt.Namespace
+	}
+	return client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+}
+
+func getOrCreateServiceAccount(client kube.ExtendedClient, opt RemoteSecretOptions) (*v1.ServiceAccount, error) {
+	if sa, err := client.CoreV1().ServiceAccounts(opt.Namespace).Get(
+		context.TODO(), opt.ServiceAccountName, metav1.GetOptions{}); err == nil {
+		return sa, nil
+	} else if !opt.CreateServiceAccount {
+		// User chose not to automatically create the service account.
+		return nil, err
+	}
+
+	if err := createServiceAccount(client, opt); err != nil {
+		return nil, err
+	}
+
+	// Return the newly created service account.
+	sa, err := client.CoreV1().ServiceAccounts(opt.Namespace).Get(
+		context.TODO(), opt.ServiceAccountName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving service account %s.%s after creating it: %v",
+			opt.ServiceAccountName, opt.Namespace, err)
+	}
+	return sa, nil
+}
+
+func createServiceAccount(client kube.ExtendedClient, opt RemoteSecretOptions) error {
+	// Create a renderer for the base installation.
+	r, err := helm.NewHelmRenderer(opt.ManifestsPath, "base", "Base", opt.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed creating Helm renderer: %v", err)
+	}
+
+	if err := r.Run(); err != nil {
+		return fmt.Errorf("failed running Helm renderer: %v", err)
+	}
+
+	values := fmt.Sprintf(`
+global:
+  istioNamespace: %s
+  istiod:
+    enableAnalysis: false
+  configValidation: false
+base:
+  enableCRDTemplates: false
+`, opt.Namespace)
+
+	yamlContent, err := r.RenderManifest(values)
+	if err != nil {
+		return fmt.Errorf("failed rendering base manifest: %v", err)
+	}
+
+	baseFile, err := writeToTempFile(yamlContent)
+	if err != nil {
+		return fmt.Errorf("failed creating manifest file: %v", err)
+	}
+
+	// Before we can apply the yaml, we have to ensure the system namespace exists.
+	if err := createNamespaceIfNotExist(client, opt.Namespace); err != nil {
+		return err
+	}
+
+	// Apply the YAML to the cluster.
+	if err := client.ApplyYAMLFiles(opt.Namespace, baseFile); err != nil {
```

Comment:
should we filter out CRDs, etc if we don't want them?

Reply:
I feel like the real fix here is to provide a mechanism for filtering them out that works. @irisdingbj is already working on something. Once that lands, we can just use that here. In the meantime, all of our installations require CRDs currently so this shouldn't be a problem.